lib/libzpool/uzfs_io.c

```diff
@@ -50,16 +87,43 @@ uzfs_write_data(zvol_state_t *zv, char *buf, uint64_t offset, uint64_t len,
 	 * Handling the case where readlen is smaller than blocksize.
 	 * This can also be avoided later for better performance.
 	 */
-	r_offset = (offset / blocksize) * blocksize;
-	r_len = ((len + blocksize - 1) / blocksize) * blocksize;
+	r_offset = P2ALIGN_TYPED(offset, blocksize, uint64_t);
+	r_len = P2ALIGN_TYPED(((offset - r_offset) + len + blocksize - 1),
+	    blocksize, uint64_t);
 
 	len_in_first_aligned_block = (blocksize - (offset - r_offset));
 
 	if (len_in_first_aligned_block > len)
 		len_in_first_aligned_block = len;
 
+start:
 	rl = zfs_range_lock(&zv->zv_range_lock, r_offset, r_len, RL_WRITER);
 
+/*
+ * Matching the cases to enter dmu_sync code during zil_commit
+ * from zvol_log_write
+ */
+	if (sync &&
+	    ((zv->zv_zilog->zl_logbias == ZFS_LOGBIAS_THROUGHPUT) ||
+	    (blocksize > ZIL_MAX_COPIED_DATA) ||
+	    (!spa_has_slogs(zv->zv_zilog->zl_spa) && len >= blocksize &&
+	    blocksize > zvol_immediate_write_sz))) {
+
+		node = kmem_alloc(sizeof (dmu_sync_node_t), KM_SLEEP);
```

Comment:
as you mentioned Vitta, it would have been better to allocate node on stack to avoid leaks and simplify code.